### PR TITLE
IGNITE-21836: KeyValueView. GetNullable produces confusing error for a PoJo when field / column nullability do not match

### PR DESCRIPTION
**Pull request by lowka**
https://issues.apache.org/jira/browse/IGNITE-21836

---

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

**Review by AMashenkov — APPROVED** [review]

File: modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientKeyValueView.java [hunk 1]
```diff
@@ -229,6 +235,11 @@ public CompletableFuture<Void> putAllAsync(@Nullable Transaction tx, Map<K, V> p
             return nullCompletedFuture();
         }
 
+        for (Entry<K, V> e : pairs.entrySet()) {
+            Objects.requireNonNull(e.getKey());
```
Comment (ygerzhedovich) [inline modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientKeyValueView.java:239]: should we add some message to help a user identify a problem?

File: modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/ValidationUtils.java [hunk 2]
```diff
@@ -95,5 +99,35 @@ private static boolean isColumnCompatible(BinaryMode colType, Class<?> cls) {
...
+
+    /**
+     * Checks whether {@code getNullable*} operation on a {@code KeyValueView} is allowed for the given value type.
+     * If the value type is not natively supported, throws an exception.
+     *
+     * @param valueType Value type.
+     */
+    public static void validateNullableOperation(Class<?> valueType) {
+        if (!Mapper.nativelySupported(valueType)) {
+            String message = format("`getNullable`* methods cannot be used when a value is not mapped to a simple type");
```
Comment (ygerzhedovich) [inline modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/ValidationUtils.java:126]: the method is utility and isn't linked to concrete classes that will be used. Maybe better to pass a method name as a parameter to form the message

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientNullsTest.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            view.putAll(null, map);
+        });
+    }
+
+    @Test
+    public void testNullableOpSingleCol() {
+        Table table = ignite().tables().table("T1");
+        KeyValueView<Integer, Boolean> view = table.keyValueView(Mapper.of(Integer.class), Mapper.of(Boolean.class));
+
+        assertEquals(NullableValue.of(null), view.getNullable(null, 1));
```
Comment (ygerzhedovich) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientNullsTest.java:99]: maybe we can make NullableValue.NULL is public

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientNullsTest.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        assertEquals(NullableValue.of(null), view.getNullable(null, 3));
+    }
+
+    private static void expectNotSupported(Executable exec) {
+        UnsupportedOperationException err = assertThrows(UnsupportedOperationException.class, exec);
+        assertEquals("`getNullable`* methods cannot be used when a value is not mapped to a simple type", err.getMessage());
+    }
+
+    private static void expectNullNotAllowed(Executable exec) {
+        NullPointerException err = assertThrows(NullPointerException.class, exec);
```
Comment (ygerzhedovich) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientNullsTest.java:133]: we have `IgniteTestUtils.assert*` methods

File: modules/table/src/main/java/org/apache/ignite/internal/table/KeyValueViewImpl.java [hunk 5]
```diff
@@ -236,7 +242,11 @@ public void putAll(@Nullable Transaction tx, Map<K, V> pairs) {
     public CompletableFuture<Void> putAllAsync(@Nullable Transaction tx, Map<K, V> pairs) {
         Objects.requireNonNull(pairs);
         for (Entry<K, V> entry : pairs.entrySet()) {
-            Objects.requireNonNull(entry.getKey());
+            K key = entry.getKey();
+            V val = entry.getValue();
+
+            Objects.requireNonNull(key);
```
Comment (ygerzhedovich) [inline modules/table/src/main/java/org/apache/ignite/internal/table/KeyValueViewImpl.java:248]: https://github.com/apache/ignite-3/pull/3714/files#r1596527346

File: modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientKeyValueView.java (hunk 1 above)
Comment (lowka) [reply modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientKeyValueView.java:239]: We should. But all assertions in this classes and their siblings do not return any message.

File: modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/ValidationUtils.java (hunk 2 above)
Comment (lowka) [reply modules/marshaller-common/src/main/java/org/apache/ignite/internal/marshaller/ValidationUtils.java:126]: This code uses already established pattern see `ClientKeyValueView::throwIfNull` its callers.

> throw new UnexpectedNullValueException("Got unexpected null value: use `getNullable` sibling method instead.");

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientNullsEmbeddedTest.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package org.apache.ignite.internal.runner.app.client;
+
+import org.apache.ignite.Ignite;
+
+/**
+ * Embedded client nulls in operations test.
+ */
+public class ItThinClientNullsEmbeddedTest extends ItThinClientNullsTest {
```
Comment (korlov42) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientNullsEmbeddedTest.java:25]: client is either _thin_ or _embedded_

**Review by korlov42 — APPROVED** [review]

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientNullsTest.java (hunk 4 above)
Comment (lowka) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientNullsTest.java]: fixed

File: modules/table/src/main/java/org/apache/ignite/internal/table/KeyValueViewImpl.java (hunk 5 above)
Comment (lowka) [reply modules/table/src/main/java/org/apache/ignite/internal/table/KeyValueViewImpl.java]: Added variable names to `requireNonNull` calls.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientNullsEmbeddedTest.java (hunk 6 above)
Comment (lowka) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItThinClientNullsEmbeddedTest.java]: Thanks. Fixed.

File: modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientKeyValueView.java [hunk 7]
```diff
@@ -94,18 +97,24 @@ public class ClientKeyValueView<K, V> extends AbstractClientView<Entry<K, V>> im
...
+        return sync(doGet(tx, key, "getNullable"));
     }
 
     /** {@inheritDoc} */
     @Override
     public CompletableFuture<V> getAsync(@Nullable Transaction tx, K key) {
-        Objects.requireNonNull(key);
+        Objects.requireNonNull(key, "key");
+
+        return doGet(tx, key, "getNullableAsync");
```
Comment (ygerzhedovich) [inline modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientKeyValueView.java:110]: should it be a 
```suggestion
        return doGet(tx, key, "getAsync");
```
?

Comment (lowka) [reply modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientKeyValueView.java:110]: No, it should not. It reports what method should be used instead.

**Review by ygerzhedovich — APPROVED** [review]

**Pull request merged**